File: travelmate-web/src/services/authService.ts
import { logger } from '../lib/utils';

const API_BASE_URL = process.env.REACT_APP_API_URL || 'http://localhost:8080/api';

export interface LoginRequest {
  email: string;
  password: string;
}

export interface LoginResponse {
  accessToken: string;
  refreshToken: string;
  expiresIn: number;
  tokenType: string;
  user: {
    id: number;
    email: string;
    nickname: string;
    fullName?: string;
    profileImageUrl?: string;
    rating: number;
    reviewCount: number;
    isEmailVerified: boolean;
    createdAt: string;
  };
}

export interface TokenResponse {
  accessToken: string;
  refreshToken: string;
  expiresIn: number;
  tokenType: string;
}

export interface OAuthLoginRequest {
  provider: 'google' | 'kakao' | 'naver';
  accessToken: string;
  deviceId?: string;
  deviceName?: string;
}

export interface OAuthCodeLoginRequest {
  provider: 'naver' | 'kakao';
  code: string;
  redirectUri: string;
  deviceId?: string;
  deviceName?: string;
}

export interface RegisterRequest {
  email: string;
  password: string;
  nickname: string;
  fullName?: string;
}

export interface RegisterResponse {
  id: number;
  email: string;
  nickname: string;
  fullName?: string;
  createdAt: string;
}

export interface CurrentUser {
  id: number;
  email: string;
  nickname: string;
  fullName?: string;
  profileImageUrl?: string;
}

class AuthService {
  // SECURITY: 토큰은 메모리에만 저장 (XSS 공격 방지)
  private accessToken: string | null = null;
  private tokenExpiresAt: number | null = null;
  private refreshPromise: Promise<TokenResponse> | null = null; // Race condition 방지
  private currentUser: CurrentUser | null = null;

  constructor() {
    // SECURITY: localStorage에서 토큰 제거 (마이그레이션)
    // 토큰은 메모리에만 저장하고, 페이지 새로고침 시 httpOnly 쿠키로 갱신
    this.migrateFromLocalStorage();

    // 저장된 사용자 ID로 사용자 정보 복원 시도
    const savedUserId = localStorage.getItem('currentUserId');
    if (savedUserId) {
      // 사용자 ID만 저장하고, 상세 정보는 토큰 갱신 시 서버에서 가져옴
      this.currentUser = { id: parseInt(savedUserId) } as CurrentUser;
    }
  }

  /**
   * localStorage에 저장된 민감한 토큰 정보 제거 (마이그레이션)
   */
  private migrateFromLocalStorage(): void {
    // 기존에 저장된 토큰 정보 모두 제거
    localStorage.removeItem('accessToken');
    localStorage.removeItem('tokenExpiresAt');
    localStorage.removeItem('authToken');
    localStorage.removeItem('refreshToken');
    localStorage.removeItem('currentUser'); // 민감한 사용자 정보도 제거
  }

  async login(request: LoginRequest): Promise<LoginResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/auth/login`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-Device-Id': this.getDeviceId(),
          'X-Device-Name': navigator.userAgent,
        },
        credentials: 'include', // httpOnly 쿠키 수신을 위해 필수
        body: JSON.stringify(request),
      });

      if (!response.ok) {
        const errorData = await response.text();
        throw new Error(`로그인 실패: ${response.status} - ${errorData}`);
      }

      const data: LoginResponse = await response.json();
      this.saveTokens(data);

      return data;
    } catch (error) {
      logger.error('Login error:', error);
      throw error;
    }
  }

  async oauthLogin(request: OAuthLoginRequest): Promise<LoginResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/auth/oauth/login`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        credentials: 'include', // httpOnly 쿠키 수신을 위해 필수
        body: JSON.stringify({
          ...request,
          deviceId: this.getDeviceId(),
          deviceName: navigator.userAgent,
        }),
      });

      if (!response.ok) {
        const errorData = await response.text();
        throw new Error(`OAuth 로그인 실패: ${response.status} - ${errorData}`);
      }

      const data: LoginResponse = await response.json();
      this.saveTokens(data);

      return data;
    } catch (error) {
      logger.error('OAuth login error:', error);
      throw error;
    }
  }

  async oauthCodeLogin(request: OAuthCodeLoginRequest): Promise<LoginResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/auth/oauth/code-login`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        credentials: 'include',
        body: JSON.stringify({
          ...request,
          deviceId: this.getDeviceId(),
          deviceName: navigator.userAgent,
        }),
      });

      if (!response.ok) {
        const errorData = await response.text();
        throw new Error(`OAuth 코드 로그인 실패: ${response.status} - ${errorData}`);
      }

      const data: LoginResponse = await response.json();
      this.saveTokens(data);

      return data;
    } catch (error) {
      logger.error('OAuth code login error:', error);
      throw error;
    }
  }

  async refreshAccessToken(): Promise<TokenResponse> {
    // 이미 갱신 중인 경우 기존 Promise 반환 (Race Condition 방지)
    if (this.refreshPromise) {
      return this.refreshPromise;
    }

    this.refreshPromise = this.doRefreshToken();

    try {
      return await this.refreshPromise;
    } finally {
      this.refreshPromise = null;
    }
  }

  private async doRefreshToken(): Promise<TokenResponse> {
    try {
      // Refresh token은 httpOnly 쿠키로 자동 전송됨
      const response = await fetch(`${API_BASE_URL}/auth/refresh`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        credentials: 'include', // httpOnly 쿠키 전송을 위해 필수
        body: JSON.stringify({
          deviceId: this.getDeviceId(),
        }),
      });

      if (!response.ok) {
        // Refresh token도 만료된 경우 로그아웃
        this.logout();
        throw new Error('세션이 만료되었습니다. 다시 로그인해주세요.');
      }

      const data: TokenResponse = await response.json();

      // SECURITY: 토큰은 메모리에만 저장
      this.accessToken = data.accessToken;
      this.tokenExpiresAt = Date.now() + data.expiresIn * 1000;

      return data;
    } catch (error) {
      logger.error('Token refresh error:', error);
      throw error;
    }
  }

  private saveTokens(data: LoginResponse): void {
    // SECURITY: 토큰은 메모리에만 저장 (XSS 공격 방지)
    this.accessToken = data.accessToken;
    this.tokenExpiresAt = Date.now() + data.expiresIn * 1000;

    // 사용자 정보 저장 (메모리)
    if (data.user) {
      this.currentUser = {
        id: data.user.id,
        email: data.user.email,
        nickname: data.user.nickname,
        fullName: data.user.fullName,
        profileImageUrl: data.user.profileImageUrl,
      };
      // SECURITY: 사용자 ID만 localStorage에 저장 (세션 복원용)
      localStorage.setItem('currentUserId', data.user.id.toString());
    }
  }

  private getDeviceId(): string {
    let deviceId = localStorage.getItem('deviceId');
    if (!deviceId) {
      deviceId = 'web-' + Math.random().toString(36).substring(2, 15);
      localStorage.setItem('deviceId', deviceId);
    }
    return deviceId;
  }

  isTokenExpired(): boolean {
    if (!this.tokenExpiresAt) return true;
    // 5분 전에 만료로 간주 (갱신 여유 시간)
    return Date.now() > this.tokenExpiresAt - 5 * 60 * 1000;
  }

  async getValidToken(): Promise<string | null> {
    if (!this.accessToken) return null;

    if (this.isTokenExpired()) {
      try {
        // Refresh token은 httpOnly 쿠키로 관리되므로 별도 체크 불필요
        await this.refreshAccessToken();
      } catch (error) {
        return null;
      }
    }

    return this.accessToken;
  }

  async register(request: RegisterRequest): Promise<RegisterResponse> {
    try {
      const response = await fetch(`${API_BASE_URL}/users/register`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(request),
      });

      if (!response.ok) {
        const errorData = await response.text();
        throw new Error(`회원가입 실패: ${response.status} - ${errorData}`);
      }

      const data: RegisterResponse = await response.json();
      return data;
    } catch (error) {
      logger.error('Register error:', error);
      throw error;
    }
  }

  async logout(logoutAll: boolean = false): Promise<void> {
    try {
      if (this.accessToken) {
        await fetch(`${API_BASE_URL}/auth/logout`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            Authorization: `Bearer ${this.accessToken}`,
          },
          credentials: 'include', // httpOnly 쿠키 전송 및 삭제를 위해 필수
          body: JSON.stringify({
            logoutAll,
          }),
        });
      }
    } catch (error) {
      logger.error('Logout error:', error);
    } finally {
      this.clearTokens();
    }
  }

  private clearTokens(): void {
    // 메모리에서 토큰 제거
    this.accessToken = null;
    this.tokenExpiresAt = null;
    this.currentUser = null;
    // localStorage에서 사용자 ID 제거
    localStorage.removeItem('currentUserId');
    // 기존 데이터 정리 (마이그레이션)
    localStorage.removeItem('accessToken');
    localStorage.removeItem('tokenExpiresAt');
    localStorage.removeItem('authToken');
    localStorage.removeItem('currentUser');
    // refreshToken은 서버에서 httpOnly 쿠키 삭제 처리
  }

  getToken(): string | null {
    return this.accessToken;
  }

  getUser(): CurrentUser | null {
    return this.currentUser;
  }

  isAuthenticated(): boolean {
    // Access Token이 있고 만료되지 않았거나, httpOnly 쿠키로 갱신 가능한 경우
    return this.accessToken !== null && !this.isTokenExpired();
  }

  hasRefreshCapability(): boolean {
    // Refresh token은 httpOnly 쿠키로 관리되므로 클라이언트에서 확인 불가
    // 토큰 갱신 시도 후 결과로 판단
    return true;
  }

  /**
   * 앱 초기화 시 세션 복원 시도
   * httpOnly 쿠키에 저장된 refresh token을 사용하여 access token 갱신
   */
  async tryRestoreSession(): Promise<boolean> {
    const savedUserId = localStorage.getItem('currentUserId');
    if (!savedUserId) {
      return false;
    }

    try {
      await this.refreshAccessToken();
      return true;
    } catch {
      // 세션 복원 실패 시 저장된 사용자 ID 제거
      localStorage.removeItem('currentUserId');
      this.currentUser = null;
      return false;
    }
  }

  async checkEmailDuplicate(email: string): Promise<boolean> {
    try {
      const response = await fetch(
        `${API_BASE_URL}/users/check-email?email=${encodeURIComponent(email)}`,
        {
          method: 'GET',
          headers: {
            'Content-Type': 'application/json',
          },
        }
      );

      if (!response.ok) {
        throw new Error(`이메일 중복 체크 실패: ${response.status}`);
      }

      const data: { exists: boolean } = await response.json();
      return data.exists;
    } catch (error) {
      logger.error('Email duplicate check error:', error);
      throw error;
    }
  }

  async checkNicknameDuplicate(nickname: string): Promise<boolean> {
    try {
      const response = await fetch(
        `${API_BASE_URL}/users/check-nickname?nickname=${encodeURIComponent(nickname)}`,
        {
          method: 'GET',
          headers: {
            'Content-Type': 'application/json',
          },
        }
      );

      if (!response.ok) {
        throw new Error(`닉네임 중복 체크 실패: ${response.status}`);
      }

      const data: { exists: boolean } = await response.json();
      return data.exists;
    } catch (error) {
      logger.error('Nickname duplicate check error:', error);
      throw error;
    }
  }

  getAuthHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };

    if (this.accessToken) {
      headers['Authorization'] = `Bearer ${this.accessToken}`;
    }

    return headers;
  }
}

export const authService = new AuthService();

File: travelmate-backend/src/main/java/com/travelmate/service/AuthService.java
package com.travelmate.service;

import com.travelmate.dto.AuthDto;
import com.travelmate.dto.UserDto;
import com.travelmate.entity.RefreshToken;
import com.travelmate.entity.User;
import com.travelmate.exception.UserException;
import com.travelmate.repository.RefreshTokenRepository;
import com.travelmate.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.client.RestTemplate;

import java.time.LocalDateTime;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Transactional
@Slf4j
public class AuthService {

    private final UserRepository userRepository;
    private final RefreshTokenRepository refreshTokenRepository;
    private final JwtService jwtService;
    private final PasswordEncoder passwordEncoder;
    private final RestTemplate restTemplate = new RestTemplate();

    @Value("${app.jwt.expiration}")
    private Long jwtExpiration;

    @Value("${app.jwt.refresh-expiration}")
    private Long refreshExpiration;

    /**
     * 로그인 처리 - Access Token + Refresh Token 발급
     */
    public AuthDto.LoginResponse login(UserDto.LoginRequest request, String deviceId, String deviceName,
                                       String ipAddress, String userAgent) {
        User user = userRepository.findByEmail(request.getEmail())
            .orElseThrow(() -> new UserException("사용자를 찾을 수 없습니다."));

        if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new UserException("비밀번호가 일치하지 않습니다.");
        }

        // Access Token 생성
        String accessToken = jwtService.generateToken(user.getId(), user.getEmail());

        // Refresh Token 생성 및 저장
        String refreshToken = createRefreshToken(user, deviceId, deviceName, ipAddress, userAgent);

        log.info("사용자 로그인: {} (deviceId: {})", user.getEmail(), deviceId);

        return AuthDto.LoginResponse.builder()
            .accessToken(accessToken)
            .refreshToken(refreshToken)
            .expiresIn(jwtExpiration / 1000)  // 초 단위
            .tokenType("Bearer")
            .user(convertToDto(user))
            .build();
    }

    /**
     * 토큰 갱신 - Refresh Token으로 새 Access Token 발급
     */
    public AuthDto.TokenResponse refreshToken(String refreshTokenStr, String deviceId) {
        RefreshToken refreshToken = refreshTokenRepository.findByToken(refreshTokenStr)
            .orElseThrow(() -> new UserException("유효하지 않은 refresh token입니다."));

        if (!refreshToken.isValid()) {
            throw new UserException("만료되었거나 취소된 refresh token입니다.");
        }

        User user = refreshToken.getUser();

        // 새 Access Token 생성
        String newAccessToken = jwtService.generateToken(user.getId(), user.getEmail());

        // Refresh Token 마지막 사용 시간 업데이트
        refreshToken.setLastUsedAt(LocalDateTime.now());
        refreshTokenRepository.save(refreshToken);

        log.info("토큰 갱신: userId={}", user.getId());

        return AuthDto.TokenResponse.builder()
            .accessToken(newAccessToken)
            .refreshToken(refreshTokenStr)  // 기존 refresh token 유지
            .expiresIn(jwtExpiration / 1000)
            .tokenType("Bearer")
            .build();
    }

    /**
     * 로그아웃 - Refresh Token 무효화
     */
    public void logout(Long userId, String refreshTokenStr, boolean logoutAll) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new UserException("사용자를 찾을 수 없습니다."));

        if (logoutAll) {
            // 모든 기기에서 로그아웃
            refreshTokenRepository.revokeAllByUser(user);
            log.info("사용자 전체 로그아웃: userId={}", userId);
        } else if (refreshTokenStr != null) {
            // 특정 토큰만 무효화
            refreshTokenRepository.findByToken(refreshTokenStr).ifPresent(token -> {
                token.setIsRevoked(true);
                refreshTokenRepository.save(token);
            });
            log.info("사용자 로그아웃: userId={}", userId);
        }
    }

    /**
     * OAuth 소셜 로그인
     */
    public AuthDto.LoginResponse oauthLogin(AuthDto.OAuthLoginRequest request, String ipAddress, String userAgent) {
        // OAuth 제공자로부터 사용자 정보 가져오기
        AuthDto.OAuthUserInfo userInfo = getOAuthUserInfo(request.getProvider(), request.getAccessToken());

        // 기존 사용자 확인 또는 새 사용자 생성
        User user = userRepository.findByEmail(userInfo.getEmail())
            .orElseGet(() -> createOAuthUser(userInfo));

        // Access Token 생성
        String accessToken = jwtService.generateToken(user.getId(), user.getEmail());

        // Refresh Token 생성 및 저장
        String refreshToken = createRefreshToken(user, request.getDeviceId(),
            request.getDeviceName(), ipAddress, userAgent);

        log.info("OAuth 로그인: {} (provider: {})", user.getEmail(), request.getProvider());

        return AuthDto.LoginResponse.builder()
            .accessToken(accessToken)
            .refreshToken(refreshToken)
            .expiresIn(jwtExpiration / 1000)
            .tokenType("Bearer")
            .user(convertToDto(user))
            .build();
    }

    /**
     * OAuth 제공자로부터 사용자 정보 가져오기
     */
    private AuthDto.OAuthUserInfo getOAuthUserInfo(String provider, String accessToken) {
        return switch (provider.toLowerCase()) {
            case "google" -> getGoogleUserInfo(accessToken);
            case "kakao" -> getKakaoUserInfo(accessToken);
            case "naver" -> getNaverUserInfo(accessToken);
            default -> throw new UserException("지원하지 않는 OAuth 제공자입니다: " + provider);
        };
    }

    private AuthDto.OAuthUserInfo getGoogleUserInfo(String accessToken) {
        // 1. 먼저 토큰 유효성 검증 (tokeninfo 엔드포인트)
        verifyGoogleToken(accessToken);

        // 2. 토큰이 유효하면 사용자 정보 조회
        String url = "https://www.googleapis.com/oauth2/v2/userinfo";
        HttpHeaders headers = new HttpHeaders();
        headers.setBearerAuth(accessToken);
        HttpEntity<String> entity = new HttpEntity<>(headers);

        try {
            ResponseEntity<Map> response = restTemplate.exchange(url, HttpMethod.GET, entity, Map.class);
            Map<String, Object> body = response.getBody();
            if (body == null) throw new UserException("Google 응답이 비어있습니다.");
            String id = (String) body.get("id");
            String email = (String) body.get("email");
            if (id == null || email == null) throw new UserException("Google 필수 정보가 없습니다.");

            return AuthDto.OAuthUserInfo.builder()
                .provider("google").providerId(id).email(email)
                .name((String) body.get("name")).profileImageUrl((String) body.get("picture")).build();
        } catch (UserException e) { throw e;
        } catch (Exception e) {
            log.error("Google OAuth 실패", e);
            throw new UserException("Google 인증에 실패했습니다.");
        }
    }

    /**
     * Google 토큰 유효성 검증
     * tokeninfo 엔드포인트를 통해 토큰의 유효성, 만료 시간, 발급자를 검증
     */
    @SuppressWarnings("unchecked")
    private void verifyGoogleToken(String accessToken) {
        String tokenInfoUrl = "https://oauth2.googleapis.com/tokeninfo?access_token=" + accessToken;

        try {
            ResponseEntity<Map> response = restTemplate.getForEntity(tokenInfoUrl, Map.class);
            Map<String, Object> tokenInfo = response.getBody();

            if (tokenInfo == null) {
                throw new UserException("Google 토큰 검증 실패: 응답이 비어있습니다.");
            }

            // 에러 응답 확인
            if (tokenInfo.containsKey("error")) {
                String error = (String) tokenInfo.get("error");
                String errorDescription = (String) tokenInfo.get("error_description");
                log.warn("Google 토큰 검증 실패: {} - {}", error, errorDescription);
                throw new UserException("Google 토큰이 유효하지 않습니다: " + error);
            }

            // 토큰 만료 시간 확인
            Object expiresInObj = tokenInfo.get("expires_in");
            if (expiresInObj != null) {
                int expiresIn;
                if (expiresInObj instanceof String) {
                    expiresIn = Integer.parseInt((String) expiresInObj);
                } else if (expiresInObj instanceof Number) {
                    expiresIn = ((Number) expiresInObj).intValue();
                } else {
                    throw new UserException("Google 토큰 만료 시간 형식 오류");
                }

                if (expiresIn <= 0) {
                    throw new UserException("Google 토큰이 만료되었습니다.");
                }
            }

            log.debug("Google 토큰 검증 성공: email={}", tokenInfo.get("email"));

        } catch (UserException e) {
            throw e;
        } catch (Exception e) {
            log.error("Google 토큰 검증 중 오류", e);
            throw new UserException("Google 토큰 검증에 실패했습니다.");
        }
    }

    @SuppressWarnings("unchecked")
    private AuthDto.OAuthUserInfo getKakaoUserInfo(String accessToken) {
        // 1. 토큰 유효성 검증
        verifyKakaoToken(accessToken);

        // 2. 사용자 정보 조회
        String url = "https://kapi.kakao.com/v2/user/me";
        HttpHeaders headers = new HttpHeaders();
        headers.setBearerAuth(accessToken);
        HttpEntity<String> entity = new HttpEntity<>(headers);

        try {
            ResponseEntity<Map> response = restTemplate.exchange(url, HttpMethod.GET, entity, Map.class);
            Map<String, Object> body = response.getBody();
            if (body == null) throw new UserException("카카오 응답이 비어있습니다.");
            Map<String, Object> account = (Map<String, Object>) body.get("kakao_account");
            if (account == null) throw new UserException("카카오 계정 정보가 없습니다.");
            String email = (String) account.get("email");
            if (email == null) throw new UserException("카카오 이메일이 없습니다.");
            Map<String, Object> profile = (Map<String, Object>) account.get("profile");

            return AuthDto.OAuthUserInfo.builder()
                .provider("kakao").providerId(String.valueOf(body.get("id"))).email(email)
                .name(profile != null ? (String) profile.get("nickname") : null)
                .profileImageUrl(profile != null ? (String) profile.get("profile_image_url") : null).build();
        } catch (UserException e) { throw e;
        } catch (Exception e) {
            log.error("Kakao OAuth 실패", e);
            throw new UserException("카카오 인증에 실패했습니다.");
        }
    }

    /**
     * 카카오 토큰 유효성 검증
     */
    @SuppressWarnings("unchecked")
    private void verifyKakaoToken(String accessToken) {
        String tokenInfoUrl = "https://kapi.kakao.com/v1/user/access_token_info";
        HttpHeaders headers = new HttpHeaders();
        headers.setBearerAuth(accessToken);
        HttpEntity<String> entity = new HttpEntity<>(headers);

        try {
            ResponseEntity<Map> response = restTemplate.exchange(tokenInfoUrl, HttpMethod.GET, entity, Map.class);
            Map<String, Object> tokenInfo = response.getBody();

            if (tokenInfo == null) {
                throw new UserException("카카오 토큰 검증 실패: 응답이 비어있습니다.");
            }

            // 토큰 만료 시간 확인 (expires_in은 초 단위)
            Object expiresInObj = tokenInfo.get("expires_in");
            if (expiresInObj != null) {
                int expiresIn = ((Number) expiresInObj).intValue();
                if (expiresIn <= 0) {
                    throw new UserException("카카오 토큰이 만료되었습니다.");
                }
            }

            log.debug("카카오 토큰 검증 성공: id={}", tokenInfo.get("id"));

        } catch (UserException e) {
            throw e;
        } catch (Exception e) {
            log.error("카카오 토큰 검증 중 오류", e);
            throw new UserException("카카오 토큰 검증에 실패했습니다.");
        }
    }

    @SuppressWarnings("unchecked")
    private AuthDto.OAuthUserInfo getNaverUserInfo(String accessToken) {
        // 1. 토큰 유효성 검증 (네이버는 별도 검증 엔드포인트가 없으므로 사용자 정보 조회로 검증)
        // 사용자 정보 조회 실패 시 토큰이 유효하지 않은 것으로 간주

        // 2. 사용자 정보 조회
        String url = "https://openapi.naver.com/v1/nid/me";
        HttpHeaders headers = new HttpHeaders();
        headers.setBearerAuth(accessToken);
        HttpEntity<String> entity = new HttpEntity<>(headers);

        try {
            ResponseEntity<Map> response = restTemplate.exchange(url, HttpMethod.GET, entity, Map.class);
            Map<String, Object> body = response.getBody();
            if (body == null) throw new UserException("네이버 응답이 비어있습니다.");

            // 응답 코드 확인 (네이버 API는 resultcode를 반환)
            String resultCode = (String) body.get("resultcode");
            if (!"00".equals(resultCode)) {
                String message = (String) body.get("message");
                log.warn("네이버 토큰 검증 실패: resultcode={}, message={}", resultCode, message);
                throw new UserException("네이버 토큰이 유효하지 않습니다: " + message);
            }

            Map<String, Object> data = (Map<String, Object>) body.get("response");
            if (data == null) throw new UserException("네이버 사용자 정보가 없습니다.");
            String id = (String) data.get("id");
            String email = (String) data.get("email");
            if (id == null || email == null) throw new UserException("네이버 필수 정보가 없습니다.");

            log.debug("네이버 토큰 검증 및 사용자 정보 조회 성공: id={}", id);

            return AuthDto.OAuthUserInfo.builder()
                .provider("naver").providerId(id).email(email)
                .name((String) data.get("name")).profileImageUrl((String) data.get("profile_image")).build();
        } catch (UserException e) { throw e;
        } catch (Exception e) {
            log.error("Naver OAuth 실패", e);
            throw new UserException("네이버 인증에 실패했습니다.");
        }
    }

    /**
     * OAuth 사용자 생성
     */
    private User createOAuthUser(AuthDto.OAuthUserInfo userInfo) {
        User user = new User();
        user.setEmail(userInfo.getEmail());
        user.setPassword(passwordEncoder.encode(UUID.randomUUID().toString()));  // 임의의 비밀번호
        user.setNickname(generateUniqueNickname(userInfo.getName()));
        user.setFullName(userInfo.getName());
        user.setProfileImageUrl(userInfo.getProfileImageUrl());
        user.setIsActive(true);
        user.setIsEmailVerified(true);  // OAuth 사용자는 이메일 인증 완료로 처리
        user.setIsLocationEnabled(false);
        user.setIsMatchingEnabled(false);

        User savedUser = userRepository.save(user);
        log.info("새로운 OAuth 사용자 생성: {} (provider: {})", savedUser.getEmail(), userInfo.getProvider());

        return savedUser;
    }

    /**
     * 유니크한 닉네임 생성
     */
    private String generateUniqueNickname(String baseName) {
        String nickname = baseName != null ? baseName : "여행자";

        // 특수문자 제거 및 길이 제한
        nickname = nickname.replaceAll("[^가-힣a-zA-Z0-9_]", "");
        if (nickname.length() > 15) {
            nickname = nickname.substring(0, 15);
        }
        if (nickname.isEmpty()) {
            nickname = "여행자";
        }

        // 중복 체크
        String originalNickname = nickname;
        int suffix = 1;
        while (userRepository.existsByNickname(nickname)) {
            nickname = originalNickname + suffix++;
        }

        return nickname;
    }

    /**
     * Refresh Token 생성 및 저장
     */
    private String createRefreshToken(User user, String deviceId, String deviceName,
                                      String ipAddress, String userAgent) {
        // 같은 기기의 이전 토큰 무효화
        if (deviceId != null) {
            refreshTokenRepository.revokeAllByUserAndDevice(user, deviceId);
        }

        String token = UUID.randomUUID().toString();

        RefreshToken refreshToken = RefreshToken.builder()
            .token(token)
            .user(user)
            .deviceId(deviceId)
            .deviceName(deviceName)
            .ipAddress(ipAddress)
            .userAgent(userAgent)
            .expiresAt(LocalDateTime.now().plusSeconds(refreshExpiration / 1000))
            .isRevoked(false)
            .build();

        refreshTokenRepository.save(refreshToken);
        return token;
    }

    private UserDto.Response convertToDto(User user) {
        return UserDto.Response.builder()
            .id(user.getId())
            .email(user.getEmail())
            .nickname(user.getNickname())
            .fullName(user.getFullName())
            .age(user.getAge())
            .gender(user.getGender())
            .profileImageUrl(user.getProfileImageUrl())
            .bio(user.getBio())
            .currentLatitude(user.getCurrentLatitude())
            .currentLongitude(user.getCurrentLongitude())
            .travelStyle(user.getTravelStyle())
            .interests(user.getInterests())
            .languages(user.getLanguages())
            .rating(user.getRating())
            .reviewCount(user.getReviewCount())
            .isEmailVerified(user.getIsEmailVerified())
            .phoneVerified(user.getPhoneVerified())
            .lastActivityAt(user.getLastActivityAt())
            .createdAt(user.getCreatedAt())
            .build();
    }
}
